examples/tests/AuctionSpec.hs

```diff
@@ -0,0 +1,219 @@
+  A.txBid p 4 `as` wallet 3
+  awaitTime (A.bidDeadline p + 1)
+  A.txHammer p
+
+-- | helper function to compute what the given wallet owns in the
+-- given state
+holdingInState :: UtxoState -> Wallet -> L.Value
+holdingInState (UtxoState m) w
+  | Just vs <- M.lookup (walletAddress w) m = utxoValueSetTotal vs
+  | otherwise = mempty
+
+successfulSingle :: TestTree
+successfulSingle =
+  testGroup
+    "Successful single-trace runs"
+    [ testCase "zero bids" $ testSucceedsFrom testInit noBids,
+      testCase "one bid" $ testSucceedsFrom testInit oneBid,
+      testCase "two bids on the same auction" $
+        testSucceedsFrom'
+          (\_ s -> testBool $ 2 == bananasIn (holdingInState s (wallet 3)))
+          testInit
+          twoBids
+    ]
+
+-- * Failing single-trace runs
+
+failingOpen :: MonadMockChain m => m ()
+failingOpen = do
+  t0 <- currentTime
+  _ <- A.txOpen (bananaParams t0) `as` wallet 2
+  return ()
+
+failingTwoBids :: MonadMockChain m => m ()
+failingTwoBids = do
+  t0 <- currentTime
+  p <- A.txOpen (bananaParams t0) `as` wallet 1
+  A.txBid p (A.minBid p) `as` wallet 2
+  A.txBid p (A.minBid p) `as` wallet 3
+  awaitTime (A.bidDeadline p + 1)
+  A.txHammer p
+
+failingSingle :: TestTree
+failingSingle =
+  testGroup
+    "Single-trace runs that are expected to fail"
+    [ testCase "opening banana auction while owning no bananas" $
+        testFailsFrom testInit failingOpen,
+      testCase "second bid lower than first" $
+        testFailsFrom'
+          isCekEvaluationFailure
+          testInit
+          failingTwoBids
+    ]
+
+-- * A (hopefully) failing double-minting attack
+
+-- Whenever we see a transaction that mints something, try to mint ont
+-- more token and pay it to the attacker. This should be ruled out by
+-- the minting policy of the thread token.
+
+-- We use the 'somehwhere' modality to try this attack on all of the
+-- successful runs from above.
+
+attack :: TxSkel -> Maybe TxSkel
+attack (TxSkel label opts cs) = case toConstraints cs of
+  is :=>: os -> (\(xi, xo) -> TxSkel label opts (xi :=>: (xo : os))) <$> modConstrs is
+    where
+      modConstrs :: [MiscConstraint] -> Maybe ([MiscConstraint], OutConstraint)
+      modConstrs [] = Nothing
+      modConstrs (Mints red pols v : is) =
+        case Value.flattenValue v of
+          [(cs, tok, n)] ->
+            Just
+              ( Mints red pols (Value.singleton cs tok (n + 1)) : is,
+                paysPK attackerPKHash (Value.singleton cs tok 1)
+              )
+          _ -> Nothing
+      modConstrs (i : is) = first (i :) <$> modConstrs is
+
+      attackerPKHash :: L.PubKeyHash
+      attackerPKHash = walletPKHash (wallet 6)
+
+tryDupTokens :: (Alternative m, MonadModalMockChain m) => m ()
+tryDupTokens = somewhere attack (noBids <|> oneBid <|> twoBids)
+
+attacks :: TestTree
+attacks =
+  testGroup
+    "Attacks"
+    [ testCase "Double-minting" $
+        testFailsFrom'
+          isCekEvaluationFailure
```

Comment:
Nice!! This is the correct way to make sure that the test fails because the minting policy (or some other script) caught the attack instead of some silly "ValueNotPreserved" or "LessThanMinAda" errors that happen before executing the script